optex/base/optex.lua

```diff
@@ -364,3 +374,242 @@ callback.add_to_callback("input_level_string",
        end
    end
 )
+--
+-- Handling of color using attributes
+--
+-- Setting of color in PDF is handled by graphics operators which change the
+-- graphics context. Colors for fills/strokes are distinguished, but apart from
+-- that, only one color is active at time and is used for all material drawn by
+-- following graphics operators, until next color is set. Each PDF content (e.g.
+-- page or form XObject) has its own graphics context, that is initialized from
+-- zero. Hence natural way of working with color (color settings are not lost
+-- by going between pages) have to be handled.
+--
+-- \TeX/ itself has no concept of color. Color has always been handled by
+-- inserting whatsits (either using `\special` for DVI or using
+-- `\pdfliteral`/`\pdfcolorstack` for PDF). It is very efficient way of doing
+-- it, \TeX/ doesn't have to know anything about colors, it just inserts the
+-- relevant literals when doing `\shipout`. However, it is also problematic in
+-- many ways:
+--
+-- \begitems
+-- * Due to the asynchronic nature of shipping out pages, the handling of
+-- initial color of page (which should be the last color of previous page) is
+-- problematic.
+-- * In \TeX/ there isn't only contiguous page content, but also headlines,
+-- footlines and inserts (including footnotes). They shouldn't just take color
+-- from what just happens to be preceding, but should be colored independently.
+-- * Whatsit nodes are not really fully transparent to all stages of \TeX/ and
+-- just turning on/off color can result in differently typeset documents. See
+-- e.g.
+--  \begitems
+--   * \url{https://tex.stackexchange.com/questions/599165/why-does-color-change-vertical-alignement-in-table-with-fixed-width/}
+--   * \url{https://tex.stackexchange.com/questions/599367/strange-vertical-spacing-after-colored-equation/}
+--  \enditems
+-- * To give a false notion of grouping with whatsits, implementations usually
+-- set color and try to restore it with `\aftergroup`, which has its own quirks,
+-- e.g:
+--\begtt
+--\setbox0=\hbox{\Red text}   % bad:  \Black is done after `\setbox`
```

Comment:
TODO:

```
Runaway argument?
\setbox 0=\hbox {\Red text} % bad: \Black is done after `\setbox ` \setbox 0=\hbox {{\Red text}} % good: \Black is done after group inside the box \endtt \enditems \_endgraf While the first two issues can be handled using `\pdfcolorstack `,\ETC.
! Paragraph ended before \_eoldefA was complete.
<to be read again>
```

Reply:
I move this part of documentation to colors.opm, the `\begtt...\endtt` problem will be solved.